<a href="https://colab.research.google.com/github/gfilippini/ExData_Ploting2/blob/master/hito_2_taller_de_datos_python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Optimización del proceso de ejecución de Programa Rural de Ministerio de Vivienda y Urbanismo**

teste
giuliano



nova alteração giuliano.filippini 21:00

## **Taller de Proyecto de Datos: Hito 2 parte 1**

####**Estudiantes**
- Militza Gascon
- Dominique Hermosilla
- Giuliano Puttini
- Gustavo Valdes


###**Objetivo de la entrega**

En este reporte se realizará el ETL y Análisis Descriptivo del proyecto "Optimización del proceso de ejecución del Programa Rural del Ministerio de Vivienda y Urbanismo", correspondiente al hito 2 del curso Taller de Proyecto de Datos.

# **Contexto del proyecto**

El Programa de Habitabilidad Rural (PHR en adelante), es un programa del Ministerio de Vivienda y Urbanismo que tiene por objetivo mejorar la calidad de vida de las familias que residen en áreas rurales, garantizando que sus viviendas cumplan con los estándares de habitabilidad establecidos por el Ministerio.

Este cuenta con tres componentes 1) Construcción de Vivienda Nueva, con las tipologías de proyectos Construcción en Sitio del Residente o Construcción de Conjunto Habitacional; 2) Mejoramiento y/o Ampliación de Vivienda Existente; 3) Mejoramiento de Entorno y Equipamiento Comunitario.

Los datos asociados a la ejecución del PHR son consolidados por parte de MINVU en un solo dataset denominado "Base PHR", que fue pedida a la institución vía solicitud de información pública en el marco de la Ley de Transparencia.

A partir de la observación inicial de los datos contenidos en el dataset se concluyó que los principales desafíos para la institución se encuentran en los tiempos de ejecución de los proyectos, razón por la cual, nuestra motivación será generar una optimización de los mismos.


El proyecto tiene como finalidad abordar las siguientes preguntas planteadas en el hito:

* ¿Se puede segmentar la información para llevar a cabo un análisis más detallado y descriptivo de cada componente del programa?
* ¿Dónde se localizan los cuellos de botella o los puntos críticos en términos de ejecución territorial?
* ¿Existen territorios o comunas particularmente desfavorecidos en la ejecución de proyectos, o, por el contrario, se han evaluado positivamente?
* ¿Es posible realizar predicciones sobre la cantidad de obras que se iniciarán y su tiempo de ejecución?

En resumen, este proyecto tiene como objetivo aprovechar la información recopilada para mejorar la planificación y ejecución del Programa de Habitabilidad Rural, garantizando que las familias rurales tengan acceso a viviendas que cumplan con los estándares de habitabilidad. Las preguntas planteadas buscan optimizar la toma de decisiones, aumentar la eficiencia en la ejecución de proyectos y identificar áreas que requieran atención especial en términos de desarrollo rural.


# **PARTE 1: ETL de los datos**

## Preparación de ambiente Python

In [ ]:
!pip uninstall pandas

Found existing installation: pandas 1.5.3
Uninstalling pandas-1.5.3:
  Would remove:
    /usr/local/lib/python3.10/dist-packages/pandas-1.5.3.dist-info/*
    /usr/local/lib/python3.10/dist-packages/pandas/*
Proceed (Y/n)? Y
  Successfully uninstalled pandas-1.5.3


In [ ]:
#Antes de ejecutar se debe reiniciar el Enterno de Ejecucion .

!pip install pandas==1.5.3

In [ ]:
!pip install -U ipywidgets ydata_profiling

In [ ]:
!pip install --upgrade typeguard typing-extensions


##Importación de librerías Pandas y Numpy, y módulo Datetime

In [ ]:
import pandas as pd
import numpy as np
import datetime as dt
#from ydata_profiling import ProfileReport

from google.colab import drive



In [ ]:
pd.__version__

'1.5.3'

##Carga y limpieza de datos

Inicialmente tiene por objetivo crear y montar el GoogleDrive que contiene la Base de Datos, y apuntar el camino donde encuentra el archivo Excel

In [ ]:

drive.mount('/content/drive',force_remount=True)
#path = '/content/drive/MyDrive/Colab_Notebooks/Taller de proyecto de datos. Grupo MINVU/Base PHR.xlsm'
path = '/content/Base PHR.xlsm'

Mounted at /content/drive


In [ ]:
#Carga de la tabla con los parametros anteriores

# utilidad para mostrar todas las columnas
pd.set_option("display.max_columns", None)


df = pd.read_excel(path
                  , sheet_name="PHR"
                  , header=[2]
                  , index_col=0
                  , na_values=["Falta","-","#N/D","00/01/1900"]
                 # , nrows=5000
                  )


In [ ]:
# Mostrar las primeras 3 filas del DataFrame. Detalle de lo que tiene el archivo
primeras_3_filas = df.head(3)

primeras_3_filas


Postula Con Proyecto Desarrollado  Año Programa N° RE Llamado  \
ID PHR                                                                 
AN10                                  No          2016            AD   
AN12                                  No          2016            AD   
AN18                                  No          2016            AD   

       Fecha Publicación DO Región Ejecución Provincia                Comuna  \
ID PHR                                                                         
AN10    2016-12-29 00:00:00      Antofagasta    El Loa  San Pedro de Atacama   
AN12    2016-12-29 00:00:00      Antofagasta    El Loa  San Pedro de Atacama   
AN18    2016-12-29 00:00:00      Antofagasta    El Loa  San Pedro de Atacama   

       Localidad  Factor Localidad Tipo de Expediente Tipología Madre  \
ID PHR                                                                  
AN10     SOCAIRE               1.6          Colectivo             CSR   
AN12     SOCAIRE               1.6          Colectivo             CSR   
AN18     SOCAIRE               1.6          Colectivo             CSR   

       Tipo de Intervención N° Expediente  m² Vivienda Nueva  \
ID PHR                                                         
AN10         Vivienda Nueva         12195                NaN   
AN12         Vivienda Nueva         12195                NaN   
AN18         Vivienda Nueva         12195                NaN   

        m² Ampliac. Cocina  m² Ampliac. Baño  m² Ampliac. Dormitorio  \
ID PHR                                                                 
AN10                   NaN               NaN                     NaN   
AN12                   NaN               NaN                     NaN   
AN18                   NaN               NaN                     NaN   

        m² Ampliac. Recinto Complem. Tipo de Recinto Complem.  \
ID PHR                                                          
AN10                             NaN                      NaN   
AN12                             NaN                      NaN   
AN18                             NaN                      NaN   

        Total m² Proyecto  CCH    CSR  Mejoram. Vivienda  Servicios Básicos  \
ID PHR                                                                        
AN10                  NaN  NaN  800.0                NaN                NaN   
AN12                  NaN  NaN  800.0                NaN                NaN   
AN18                  NaN  NaN  800.0                NaN                NaN   

        Acond. Térmico  Eficiencia Energ.  Cocina  Baño  Tercer Dormit.  \
ID PHR                                                                    
AN10               NaN                NaN     NaN   NaN             NaN   
AN12               NaN                NaN     NaN   NaN           192.0   
AN18               NaN                NaN     NaN   NaN             NaN   

        Recinto Complem.  Inmediato  Equip. Comunit.  Mejoram. Terreno  \
ID PHR                                                                   
AN10               192.0       40.0              NaN             130.0   
AN12                 NaN       40.0              NaN             130.0   
AN18               192.0       40.0              NaN             130.0   

        Solución Sanit. y/o Dotación Energía  Arq. Local  \
ID PHR                                                     
AN10                                    50.0       100.0   
AN12                                    50.0       100.0   
AN18                                    50.0       100.0   

        Discapacidad / Mov. Reducida  Regulariz.  Equip. y Espacio Público  \
ID PHR                                                                       
AN10                             NaN         NaN                       NaN   
AN12                             NaN         NaN                       NaN   
AN18                             NaN         NaN                       NaN   

        Ahorro Eximido (Aporte MINVU

Con respecto a los datos, estos se proporcionaron en un archivo Excel que contiene un total de 44,276 registros distribuidos en 153 columnas. La columna "ID PHR" identifica a los beneficiarios del programa, quienes tienen la capacidad de presentar solicitudes de proyectos de forma individual o colectiva, y pueden beneficiarse en cualquiera de las tres líneas de atención del Programa de Habitabilidad Rural.

La modalidad de las solicitudes se puede determinar a través del análisis de las columnas "Tipo de Expediente" y "N° de Expediente". En las secciones posteriores, se detallarán los pasos llevados a cabo durante la carga, procesamiento y limpieza de estos datos en un entorno de Jupyter Notebook en Google Colab.


###Limpieza y corrección de fechas

Dado que los datos en su mayoria son ingresados manualmente, las fechas que contiene el archivo Excel presentan diversos formatos (como aaaa-mm-dd, mm/dd/yyyy, mm/dd/yy, entre otros).

También se identificaron casos en los que, debido a la falta de columnas y/o la imprevisibilidad de nuevas etapas que debían ser registradas, estas se agregaron a la tabla con un separador "+".

En los casos necesarios, se trabajaron las fechas para ajustarlas al formato datetime.

Las variables afectadas por esta situacion son las siguientes:

- Fecha Publicación DO
- Fecha RE Incr. Obra
- Fecha RE Complem.
- Fecha RE Aprobación Proyecto DITEC
- Fecha Término de Obra DITEC

In [ ]:
df['Fecha Publicación DO']=df.loc[:,'Fecha Publicación DO'].apply(pd.to_datetime,dayfirst=True)
fechas = df.loc[:,['Fecha RE Incr. Obra','Fecha RE Complem.','Fecha RE Aprobación Proyecto DITEC','Fecha Término de Obra  DITEC']]
mask = fechas['Fecha RE Incr. Obra'].str.find("+")>0
fechas[['fecha_incr_1','fecha_incr_2']] = fechas.loc[:,'Fecha RE Incr. Obra'].str.split("+",expand=True,).apply(lambda x: pd.to_datetime(x,dayfirst=True))
f_incr_aux = fechas.loc[~mask,'Fecha RE Incr. Obra'].apply(lambda x: pd.to_datetime(x,dayfirst=True))
fechas = fechas.join(f_incr_aux,how='left',rsuffix='_')
fechas['fecha_incr_1']=np.where(mask,fechas['fecha_incr_1'],fechas['Fecha RE Incr. Obra_'])
mask = fechas['Fecha RE Complem.'].str.find("+")>0
fechas[['fecha_complem_1','fecha_complem_2','fecha_complem_3']] = fechas.loc[:,'Fecha RE Complem.'].str.split("+",expand=True).apply(lambda x: x.str.strip()).apply(lambda x: pd.to_datetime(x,dayfirst=True))
f_compl_aux = fechas.loc[~mask,'Fecha RE Complem.'].apply(lambda x: pd.to_datetime(x,dayfirst=True))
fechas = fechas.join(f_compl_aux,how='left',rsuffix="_")
fechas['fecha_complem_1'] = np.where(mask,fechas['fecha_complem_1'],fechas['Fecha RE Complem._'])
fechas['fecha_aprob_proyecto_DITEC'] = fechas['Fecha RE Aprobación Proyecto DITEC'].apply(lambda x: x if isinstance(x,dt.datetime) else np.nan)
fechas['fecha_term_DITEC']=fechas.loc[:,'Fecha Término de Obra  DITEC'].apply(lambda x: x if isinstance(x,dt.datetime) else np.nan)
fechas.head()

Fecha RE Incr. Obra Fecha RE Complem.  \
ID PHR                                         
AN10                   NaN               NaN   
AN12                   NaN               NaN   
AN18                   NaN               NaN   
AN20                   NaN               NaN   
AN26                   NaN               NaN   

       Fecha RE Aprobación Proyecto DITEC Fecha Término de Obra  DITEC  \
ID PHR                                                                   
AN10                             00:00:00                     00:00:00   
AN12                             00:00:00                     00:00:00   
AN18                             00:00:00                     00:00:00   
AN20                             00:00:00                     00:00:00   
AN26                             00:00:00                     00:00:00   

       fecha_incr_1 fecha_incr_2 Fecha RE Incr. Obra_ fecha_complem_1  \
ID PHR                                                                  
AN10            NaT          NaT                  NaT             NaT   
AN12            NaT          NaT                  NaT             NaT   
AN18            NaT          NaT                  NaT             NaT   
AN20            NaT          NaT                  NaT             NaT   
AN26            NaT          NaT                  NaT             NaT   

       fecha_complem_2 fecha_complem_3 Fecha RE Complem._  \
ID PHR                                                      
AN10               NaT             NaT                NaT   
AN12               NaT             NaT                NaT   
AN18               NaT             NaT                NaT   
AN20               NaT             NaT                NaT   
AN26               NaT             NaT                NaT   

       fecha_aprob_proyecto_DITEC fecha_term_DITEC  
ID PHR                                              
AN10                          NaT              NaT  
AN12                          NaT              NaT  
AN18                          NaT              NaT  
AN20                          NaT              NaT  
AN26                          NaT              NaT

In [ ]:
#validación de fechas trabajadas
fechas.loc[['AN1','CO611','LR225','LA3114','AY186','LA3551'],:]

Fecha RE Incr. Obra                     Fecha RE Complem.  \
ID PHR                                                                  
AN1         2020-08-06 00:00:00                                   NaN   
CO611                       NaN                 10-08-2021 + 15-11-21   
LR225   11-06-2021 + 30-12-2021                                   NaN   
LA3114                      NaN                   2020-12-29 00:00:00   
AY186                       NaN                   2020-03-23 00:00:00   
LA3551                      NaN  02-06-2021 + 23-11-2022 + 05-12-2022   

       Fecha RE Aprobación Proyecto DITEC Fecha Término de Obra  DITEC  \
ID PHR                                                                   
AN1                   2018-07-10 00:00:00          2022-06-21 00:00:00   
CO611                 2021-12-02 00:00:00                     00:00:00   
LR225                 2018-07-03 00:00:00          2020-12-21 00:00:00   
LA3114                2021-02-12 00:00:00                     00:00:00   
AY186                            00:00:00                     00:00:00   
LA3551                2021-06-17 00:00:00                     00:00:00   

       fecha_incr_1 fecha_incr_2 Fecha RE Incr. Obra_ fecha_complem_1  \
ID PHR                                                                  
AN1      2020-08-06          NaT           2020-08-06             NaT   
CO611           NaT          NaT                  NaT      2021-08-10   
LR225    2021-06-11   2021-12-30                  NaT             NaT   
LA3114          NaT          NaT                  NaT      2020-12-29   
AY186           NaT          NaT                  NaT      2020-03-23   
LA3551          NaT          NaT                  NaT      2021-06-02   

       fecha_complem_2 fecha_complem_3 Fecha RE Complem._  \
ID PHR                                                      
AN1                NaT             NaT                NaT   
CO611       2021-11-15             NaT                NaT   
LR225              NaT             NaT                NaT   
LA3114             NaT             NaT         2020-12-29   
AY186              NaT             NaT         2020-03-23   
LA3551      2022-11-23      2022-12-05                NaT   

       fecha_aprob_proyecto_DITEC fecha_term_DITEC  
ID PHR                                              
AN1                    2018-07-10       2022-06-21  
CO611                  2021-12-02              NaT  
LR225                  2018-07-03       2020-12-21  
LA3114                 2021-02-12              NaT  
AY186                         NaT              NaT  
LA3551                 2021-06-17              NaT

In [ ]:
#inserción de fechas en la base de datos:
lista = ['fecha_incr_1','fecha_incr_2','fecha_complem_1','fecha_complem_2','fecha_complem_3','fecha_aprob_proyecto_DITEC','fecha_term_DITEC']
df = df.join(fechas[lista],how='left', rsuffix='_fechas')

###Selección de variables de interés

A pesar de la gran cantidad de variables en la base de datos PHR, a priori, se observa que muchas características son irrelevantes para el estudio propuesto. Esta intuición ha sido confirmada por la experta de la institución que actúa como contarparte, quien nos ha explicitado que los nudos críticos del proceso se encuentran en torno de las fechas de los procesos y los datos de tipo financiero.

Adicionalmente, la base cuenta con diversas informaciones de números de requerimiento interno (RE), que identifican los procesos o solicitudes ingresadas. Estos datos no se considerarán por ser de tipo categórico nominal. Como se explicó anteriormente, la variable `N° de Expediente` es relevante para identificar grupos de solicitudes de modalidad colectiva.

Por otro lado, la tabla cuenta con diversas informaciones de detalles de m² de obras, o los detalles de valores de cada etapa de la obra, pero no han sido confirmados por la experta como datos imprescindibles.

Así, las características a ser trabajadas pueden ser reducidas en gran cantidad, mantenendo solamente las informaciones relevantes para análisis.

El grupo de informaciones _Homologación_ _de_ _Subsidio_ _(RE 833/2023)_ conforman una nueva etapa dentro del proceso, pero dado la recién incorporación de este proceso dentro del Programa Rural, no se cuenta con gran poblamiento de datos. Así, ese grupo de informaciones quedarán excluídas de los análisis.

In [ ]:
# Creación de un diccionário de datos:

dic = {'Identificación del Subsidio':['Postula Con Proyecto Desarrollado',	'Año Programa',	'N° RE Llamado',	'Fecha Publicación DO']
       ,'Datos Generales':['Región Ejecución',	'Provincia',	'Comuna',	'Localidad',	'Factor Localidad',	'Tipo de Expediente',	'Tipología Madre',	'Tipo de Intervención',	'N° Expediente']
       ,'Dimensión':['m² Vivienda Nueva',	'm² Ampliac. Cocina',	'm² Ampliac. Baño',	'm² Ampliac. Dormitorio',	'm² Ampliac. Recinto Complem.',	'Tipo de Recinto Complem.',	'Total m² Proyecto']
       ,'Vivienda Nueva': ['CCH', 'CSR']
       ,'Mejoramiento': ['Mejoram. Vivienda',	'Servicios Básicos',	'Acond. Térmico',	'Eficiencia Energ.']
       ,'Ampliación':['Cocina',	'Baño',	'Tercer Dormit.',	'Recinto Complem.']
       ,'Entorno':['Inmediato',	'Equip. Comunit.']
       ,'Complementario':['Mejoram. Terreno',	'Solución Sanit. y/o Dotación Energía',	'Arq. Local',	'Discapacidad / Mov. Reducida',	'Regulariz.',	'Equip. y Espacio Público']
       ,'Ahorro':['Ahorro Eximido (Aporte MINVU)',	'Ahorro Familia',	'Otros Aportes Adicion.',	'Origen Aportes Adicionales']
       ,'Total Obra':['Total Asignado Proyecto MINVU',	'Total Presup. Proyecto']
       ,'Assistencia Técnica':['AT',	'FTO',	'CBR', 'Total AT + FTO + CBR']
       ,'Incremento Regular Obra':['Monto Incr. Obra',	'Año Imputación Incr. Obra',	'N° RE Incr. Obra',	'Fecha RE Incr. Obra',	'Concepto Incremento Obra']
       ,'Incremento Regular AT + FTO + CBR':['Monto Incr. AT + FTO + CBR',	'Año Imputación Incr. AT + FTO + CBR',	'N° RE Incr. AT + FTO + CBR',	'Fecha RE Incr. AT + FTO + CBR',	'Concepto Incremento AT + FTO + CBR', 'Total Incremento']
       ,'COMPLEMENTARIO SEREMI (RE 6042/2016)':['Complem. Mejoram. Terreno',	'Complem. Solución Sanit. y/o Dotación Energía',	'Complem. Arq. Local',	'Complem. Discapacidad / Mov. Reducida',	'Complem. Regulariz.',	'Complem. Equip. y Espacio Público',	'Complem. AT',	'Complem. FTO',	'Año Imputación Complem.',	'N° RE Complem.',	'Fecha RE Complem.']
       ,'Total Complem.':['Total Complem.']
       ,'INCREMENTO ALZA MATERIALES (RE 1448/2021)':['Monto 1er Incr. Alza Materiales',	'Año Imputación 1er Incr. Alza Materiales',	'N° RE 1er Incr. Alza Materiales',	'Fecha RE 1er Incr. Alza Materiales',	'Monto 2do Incr. Alza Materiales',	'Año Imputación 2do Incr. Alza Materiales',	'N° RE 2do Incr. Alza Materiales',	'Fecha RE 2do Incr. Alza Materiales',	'Monto 3er Incr. Alza Materiales',	'Año Imputación 3er Incr. Alza Materiales',	'N° RE 3er Incr. Alza Materiales',	'Fecha RE 3er Incr. Alza Materiales', 'Total Alza Materiales']
       ,'HOMOLOGACIÓN DE SUBSIDIO (RE 833/2023)':{'Vivienda Nueva':['Homolog. CCH',	'Homolog. CSR']
                                                  ,'Mejoramiento':['Homolog. Mejoram. Vivienda', 'Homolog. Servicios Básicos',	'Homolog. Acond. Térmico',	'Homolog. Eficiencia Energ.']
                                                  ,'Ampliación':['Homolog. Cocina',	'Homolog. Baño',	'Homolog. Tercer Dormitorio',	'Homolog. Recinto Complem.']
                                                  ,'Entorno':['Homolog. Inmediato',	'Homolog. Equip. Comunit.']
                                                  ,'Complementario':['Homolog. Mejoram. Terreno',	'Homolog. Solución Sanit. y/o Dotación Energía',	'Homolog. Arq. Local',	'Homolog. Discapacidad / Mov. Reducida',	'Homolog. Regulariz.',	'Homolog. Equip. y Espacio Público']
                                                  ,'AT y FTO':['Homolog. AT',	'Homolog. FTO']
                                                  ,'Homologacion':['Año Imputación Homolog.',	'N° RE Homolog.', 	'Fecha RE Homolog.']
                                                  ,'Total Homolog.':['Total Homolog']
                                                  }
       ,'Alberge Transitorio':['Monto 1er Albergue Tansitorio',	'Año Imputación 1er Albergue Transitorio',	'RE 1er Albergue Transitorio',	'Fecha RE 1er Albergue Transitorio',	'Monto 2do Albergue Tansitorio',	'Año Imputación 2do Albergue Transitorio',	'RE 2do Albergue Transitorio',	'Fecha RE 2do Albergue Transitorio',	'Monto 3er Albergue Tansitorio',	'Año Imputación 3er Albergue Transitorio',	'RE 3er Albergue Transitorio',	'Fecha RE 3er Albergue Transitorio', 'Total Albergue Transitorio']
       ,'TOTALES':['Total Aporte MINVU Original',	'Total Aporte MINVU Final (incluye incrementos)',	'Total Disponible Proyecto (incl. ahorro familia)']
       ,'Ejecución de Obras':['Proyección Inicio Obra (año/semestre)',	'Incremento Requerido para Iniciar Obras',	'N° RE Aprobación Proyecto',	'Fecha RE Aprobación Proyecto',	'N° RE Aprobación Proyecto DITEC',	'Fecha RE Aprobación Proyecto DITEC',	'Fecha Acta Entrega Terreno',	'Fecha Acta Entrega Terreno DITEC',	'Nombre Constructora',	'Nombre Constructora DITEC',	'Fecha Recepción SERVIU',	'Fecha Recepción DOM',	'Fecha Término de Obra',	'Comparar Actualización Término de Obra DITEC',	'Fecha Término de Obra  DITEC',	'Año Término',	'Cumple Término PEH']
       ,'ESTADO PROYECTO':['% Avance Físico',	'ESTADO PROYECTO',	'% Avance Físico DITEC',	'ESTADO PROYECTO DITEC',	'SUBESTADO PROYECTO DITEC',	'% Corregido',	'Estado beneficiario DITEC',	'N° RE Renuncia / Eliminación',	'Fecha RE Renuncia / Eliminación',	'Observación DPH',	'Observación Región']
       }

In [ ]:
keys = ['Identificación del Subsidio','Datos Generales','Ahorro','Total Obra','Total Complem.', 'TOTALES', 'Ejecución de Obras', 'ESTADO PROYECTO']
lista=[]
for key in keys:
  for item in dic[key]:
    lista.append(item)
fechas = ['fecha_incr_1','fecha_incr_2','fecha_complem_1','fecha_complem_2','fecha_complem_3','fecha_aprob_proyecto_DITEC','fecha_term_DITEC']
for item in fechas:
  lista.append(item)
otros = ['Monto Incr. Obra', 'Concepto Incremento Obra', 'Total Incremento', 'Fecha RE 1er Incr. Alza Materiales', 'Fecha RE 2do Incr. Alza Materiales', 'Fecha RE 3er Incr. Alza Materiales', 'Total Alza Materiales', 'RE 1er Albergue Transitorio',	'Fecha RE 1er Albergue Transitorio', 'Fecha RE 2do Albergue Transitorio', 'Fecha RE 3er Albergue Transitorio', 'Total Albergue Transitorio']
for item in otros:
  lista.append(item)
retirar = ['Año Programa','Comuna', 'Localidad', 'Factor Localidad', 'N° RE Aprobación Proyecto', 'N° RE Aprobación Proyecto DITEC','Comparar Actualización Término de Obra DITEC', 'Año Término']
for item in retirar:
  lista.remove(item)
lista

df_f = df[lista]

df_f.head(3)


Postula Con Proyecto Desarrollado N° RE Llamado Fecha Publicación DO  \
ID PHR                                                                        
AN10                                  No            AD           2016-12-29   
AN12                                  No            AD           2016-12-29   
AN18                                  No            AD           2016-12-29   

       Región Ejecución Provincia Tipo de Expediente Tipología Madre  \
ID PHR                                                                 
AN10        Antofagasta    El Loa          Colectivo             CSR   
AN12        Antofagasta    El Loa          Colectivo             CSR   
AN18        Antofagasta    El Loa          Colectivo             CSR   

       Tipo de Intervención N° Expediente  Ahorro Eximido (Aporte MINVU)  \
ID PHR                                                                     
AN10         Vivienda Nueva         12195                            NaN   
AN12         Vivienda Nueva         12195                            NaN   
AN18         Vivienda Nueva         12195                            NaN   

        Ahorro Familia  Otros Aportes Adicion. Origen Aportes Adicionales  \
ID PHR                                                                      
AN10              10.0                     NaN                        NaN   
AN12              10.0                     NaN                        NaN   
AN18              10.0                     NaN                        NaN   

        Total Asignado Proyecto MINVU  Total Presup. Proyecto  Total Complem.  \
ID PHR                                                                          
AN10                           1312.0                  1322.0             0.0   
AN12                           1312.0                  1322.0             0.0   
AN18                           1312.0                  1322.0             0.0   

        Total Aporte MINVU Original  \
ID PHR                                
AN10                         1416.6   
AN12                         1416.6   
AN18                         1416.6   

        Total Aporte MINVU Final (incluye incrementos)  \
ID PHR                                                   
AN10                                            1416.6   
AN12                                            1416.6   
AN18                                            1416.6   

        Total Disponible Proyecto (incl. ahorro familia)  \
ID PHR                                                     
AN10                                              1426.6   
AN12                                              1426.6   
AN18                                              1426.6   

       Proyección Inicio Obra (año/semestre)  \
ID PHR                                         
AN10                                     NaN   
AN12                                     NaN   
AN18                                     NaN   

       Incremento Requerido para Iniciar Obras Fecha RE Aprobación Proyecto  \
ID PHR                                                                        
AN10                                       NaN                          NaT   
AN12                                       NaN                          NaT   
AN18                                       NaN                          NaT   

       Fecha RE Aprobación Proyecto DITEC Fecha Acta Entrega Terreno  \
ID PHR                                                                 
AN10                             00:00:00                        NaT   
AN12                             00:00:00                        NaT   
AN18                             00:00:00                        NaT   

       Fecha Acta Entrega Terreno DITEC Nombre Constructora  \
ID PHR                                                        
AN10                           00:00:00          AtacamaSun   
AN12                           00:00:00          AtacamaSun   
AN18                           00:00:00

###Exclusión de informaciones

En algunos casos, las informaciones son incoherentes o erróneas, sea por problemas sistémicos, o por la extración de los datos del sistema origen.

Cuando no se puede averiguar la corrección de los datos desde la fuente, puede ser mejor excluirlos de los análisis para no inducir variabilidad proveniente de problemas sistémicos ajnos al problema de estudio.

En este proyecto fueron identificados casos en que la fecha de conclusión de obras estaban marcados como **anteriores** a la fecha de inicio de la postulación. Así, esos casos fueron excluidos del análisis.

In [ ]:
dif = (df_f['Fecha Término de Obra'] - df_f['Fecha Publicación DO']).dt.days / 365
mask = dif < 0
#código abajo sirve para averiguar quienes son - no necesário para análisis
#df_f[mask]

#ese código filtra los datos incoerentes
df_f = df_f[~mask]
print(df_f.shape)
#en los casos en que no hay Fecha Publicación, sustituir por el Año del Programa
dif = df_f['Fecha Término de Obra'].dt.year - df_f['Año Programa']
mask = dif < 0
df_f = df_f[~mask]
print(df_f.shape)

(44265, 64)
(44264, 64)


###Creación de variables adicionales

Grande parte de los análisis estadísticos consisten en crear nuevas variables basadas en los datos.

Para este proyecto, el interés principal está en la duración de tiempo toma desde la entrada del proyecto hasta la fecha de término. Para el Ministerio, eso cuantifica sus indicadores de performance y miden el retraso del proceso de entrega de los beneficios.

Otra variable de interés tiene que a ver con la cantidad de etapas que cada proyecto tiene desde el comienzo hasta la publicación. Se espera que, etapas adicionales evidencian complejidades para el proceso y pueden consistir en fallas o problemas en las entregas dentro del plazo esperado.

**Duración del proceso hasta conclusión (en anõs)**

In [ ]:
#En los casos que la Fecha Publicación DO no esta cargada en la tabla, se sugiere utilizar el Año Programa para cálculo
#de la duración
mask=df_f['Fecha Publicación DO'].isnull()
df_f[mask]
df_f['Duracion (años)'] = np.where(mask
                                   ,df_f['Fecha Término de Obra'].dt.year - df_f['Año Programa']
                                   ,(df_f['Fecha Término de Obra'] - df_f['Fecha Publicación DO']).dt.days / 365
                                  )

# Estadísticas descriptivas en años
estadisticas_duracion_anual = df_f['Duracion (años)'].describe()

# Crear un DataFrame para mostrar las estadísticas en años
df_estadisticas_duracion_anual = pd.DataFrame(estadisticas_duracion_anual)
df_estadisticas_duracion_anual.reset_index(inplace=True)
df_estadisticas_duracion_anual.columns = ['Métrica', 'Duración (años)']

# Mostrar las estadísticas descriptivas en años
print(df_estadisticas_duracion_anual)

  Métrica  Duración (años)
0   count     12219.000000
1    mean         2.806573
2     std         1.187633
3     min         0.000000
4     25%         1.928767
5     50%         2.616438
6     75%         3.671233
7     max         6.558904


# **PARTE 2: Análisis exploratorio de los datos**

Se presenta un análisis inicial de la base de datos que, tras haber pasado por el proceso de Extracción, Transformación y Carga (ETL), ahora consta de 44,264 registros y 64 variables.

Con respecto a los datos nulos, es importante destacar que las variables más relevantes para este estudio se encuentran completas. Esto incluye la identificación del tipo de subsidio, abarcando variables como "Tipo de Intervención," "Tipología Madre" y "Tipo de Expediente."

Además, otras variables de gran relevancia están relacionadas con la ubicación geográfica de los subsidios, tales como "Región de Ejecución," "Provincia," "Comuna" y "Localidad." Estas variables son críticas para comprender la distribución geográfica de los subsidios y su impacto en diferentes áreas.

Finalmente, las variables que informan sobre el estado del proyecto y las fechas que marcan los hitos del proceso de cada subsidio son igualmente esenciales. Estas variables permitirán al proyecto llevar a cabo análisis en profundidad y responder de manera efectiva a las preguntas planteadas, ya que proporcionarán información detallada sobre el progreso y desarrollo de los proyectos de subsidio.







In [ ]:
df_f.info()

<class 'pandas.core.frame.DataFrame'>
Index: 44265 entries, AN10 to VA1713
Data columns (total 64 columns):
 #   Column                                            Non-Null Count  Dtype         
---  ------                                            --------------  -----         
 0   Postula Con Proyecto Desarrollado                 44265 non-null  object        
 1   Año Programa                                      44265 non-null  int64         
 2   N° RE Llamado                                     44265 non-null  object        
 3   Fecha Publicación DO                              39094 non-null  datetime64[ns]
 4   Región Ejecución                                  44265 non-null  object        
 5   Provincia                                         44265 non-null  object        
 6   Tipo de Expediente                                44265 non-null  object        
 7   Tipología Madre                                   44265 non-null  object        
 8   Tipo de Intervención       

## Análisis de los componentes del Programa de Habitabilidad Rural




En el análisis de los datos, se observa que la variable "Tipo de Intervención" se divide en dos categorías principales: "Vivienda Nueva" con una frecuencia de 23,094, representando el 52.16% de los casos, y "Mejoramiento" con una frecuencia de 21,182, que equivale al 47.84% de los casos. Esto sugiere que la mayoría de los registros corresponden a proyectos de "Vivienda Nueva," mientras que una proporción significativa también se refiere a proyectos de "Mejoramiento." Estos hallazgos son fundamentales para comprender la distribución de los tipos de intervención en la base de datos.


In [ ]:
# Calcular frecuencias
frecuencias_tipo_intervencion = df_f['Tipo de Intervención'].value_counts()

# Calcular proporciones
proporciones_tipo_intervencion = df_f['Tipo de Intervención'].value_counts(normalize=True).round(4)*100

# Crear un DataFrame para mostrar las frecuencias y proporciones
tabla_tipo_intervencion = pd.DataFrame({'Frecuencia': frecuencias_tipo_intervencion, 'Proporción': proporciones_tipo_intervencion})

# Mostrar la tabla
print(tabla_tipo_intervencion)

                Frecuencia  Proporción
Vivienda Nueva       23093       52.17
Mejoramiento         21171       47.83


Respecto de la distribución de proyectos según la "Tipología Madre" y el "Tipo de Expediente". En la categoría "Tipología Madre" denominada "CCH" (Construccion de Conjunto Habitacional) se observa un total de 5,816 proyectos. La gran mayoría de estos proyectos son de tipo "Colectivo" (5,808), con solo un número muy reducido de proyectos "Individual" (8).

En la categoría "CSR" (Construccion en Sitio del Residente) se encuentran 17,277 proyectos. La mayoría de ellos son de tipo "Colectivo" (11,329), seguidos por una cantidad significativa de proyectos "Individual" (5,948).

La categoría "MAVE" (mejoramiento y/o Ampliacion de Vivienda Existente)  cuenta con un total de 20,172 proyectos. La gran mayoría son "Colectivos" (16,760), seguidos por proyectos "Individual" (3,412).

En la categoría "MEEC"v (Mejoramiento del Entorno y Equipamiento Comunitario)  se registra un total de 999 proyectos, y todos ellos son de tipo "Colectivo." ya que la normativa no permite proyectos de tipo "Individual" en esta modalidad.

En resumen, este análisis revela la distribución de proyectos según su "Tipología Madre" y el "Tipo de Expediente." La mayoría de los proyectos en estas categorías son de tipo "Colectivo," con un número muy limitado de proyectos "Individual." Esto proporciona información útil para comprender la estructura de los proyectos.

In [ ]:
pd.crosstab(df_f['Tipología Madre'],df_f['Tipo de Expediente'],margins=True)

Tipo de Expediente  COLECTIVO  INDIVIDUAL    All
Tipología Madre                                 
CCH                      5808           8   5816
CSR                     11329        5948  17277
MAVE                    16760        3412  20172
MEEC                      999           0    999
All                     34896        9368  44264

El análisis de los datos revela una variada distribución en cuanto al estado de los subsidios en la base de datos. La categoría "Obra Terminada" cuenta con la frecuencia más alta, representando el 27.60% de los casos, seguida de cerca por "Obra en Ejecución," que abarca el 25.41%. "Proyecto Aprobado" y "Sin Proyecto Ingresado" constituyen el 15.38% y el 15.23% de los registros, respectivamente. Mientras tanto, "Proyecto en Evaluación" comprende el 12.74% de los casos, y "Eliminado/Renunciado" representa el 3.64%. Estos resultados indican que la mayoría de los subsidios en la base de datos se encuentran en etapas de "Obra Terminada" o "Obra en Ejecución," mientras que las categorías "Proyecto Aprobado" y "Sin Proyecto Ingresado" también tienen una representación significativa. Las categorías restantes, aunque menos frecuentes, también aportan información valiosa sobre el estado de los proyectos de subsidio.

In [ ]:
# Calcular frecuencias
frecuencias_estado_proyecto = df_f['ESTADO PROYECTO'].value_counts()

# Calcular proporciones
proporciones_estado_proyecto = df_f['ESTADO PROYECTO'].value_counts(normalize=True).round(4)*100

# Crear un DataFrame para mostrar las frecuencias y proporciones
tabla_estado_proyecto = pd.DataFrame({'Subsidios': frecuencias_estado_proyecto, 'Proporción': proporciones_estado_proyecto})

# Mostrar la tabla
print(tabla_estado_proyecto)

                        Subsidios  Proporción
OBRA TERMINADA              12217       27.60
OBRA EN EJECUCIÓN           11249       25.41
PROYECTO APROBADO            6806       15.38
SIN PROYECTO INGRESADO       6743       15.23
PROYECTO EN EVALUACIÓN       5638       12.74
ELIMINADO/RENUNCIADO         1611        3.64



En cuanto al Estado de los Proyecto vistos por tipos de proyectos, se observa que, "MAVE" (Mejoramiento y/o Ampliacionde Vivienda Existente) lidera en la mayoría de los estados, excepto en el estado "Eliminado/renunciado", donde "CSR" (Construccion en Sitio del Residente) es más prominente. Los tipos "CCH" (Construccion de Conjunto Habitacional) y "CSR" (Construccion en Sitio del Residente) también tienen una presencia considerable en varios estados, mientras que "MEEC" es menos común en general.


In [ ]:
pd.crosstab(df_f['ESTADO PROYECTO'],df_f['Tipología Madre'],margins=True)

Tipología Madre          CCH    CSR   MAVE  MEEC    All
ESTADO PROYECTO                                        
ELIMINADO/RENUNCIADO     166    526    906    13   1611
OBRA EN EJECUCIÓN       1458   4493   5298     0  11249
OBRA TERMINADA          1162   4341   6641    85  12229
PROYECTO APROBADO         27   3666   3046    67   6806
PROYECTO EN EVALUACIÓN  1106   2286   1920   326   5638
SIN PROYECTO INGRESADO  1897   1966   2372   508   6743
All                     5816  17278  20183   999  44276

## Análisis Territorial del Programa de Habitabilidad Rural


Este resultado refleja la distribución de los subsidios según la variable "Región de Ejecución". La región de La Araucanía tiene la mayor cantidad de registros con 11,221 beneficiarios, seguida de Maule con 7,667 y Ñuble con 6,275. En contraste, las regiones de Arica y Antofagasta tienen la menor representación con 34 y 75 casos, respectivamente. Estos datos indican que la mayoría de los registros provienen de las regiones de Araucanía, Maule y Ñuble, lo que puede ser relevante para analizar la distribución geográfica de los proyectos en la base de datos.

In [ ]:
subsidios_region = df_f['Región Ejecución'].value_counts()
print(subsidios_region)

Araucanía        11221
Maule             7667
Ñuble             6275
Los Lagos         3391
Los Ríos          3348
Biobío            2582
O'higgins         2200
Metropolitana     2173
Valparaíso        2109
Coquimbo          1925
Aysén              803
Atacama            276
Magallanes          93
Tarapacá            92
Antofagasta         75
Arica               34
Name: Región Ejecución, dtype: int64



Este resultado proporciona una visión detallada de la distribución de los proyectos en diferentes estados en función de la "Región de Ejecución". En la región de Araucanía, encontramos el mayor número de proyectos, con 11,221 casos en total. La mayoría de estos proyectos se encuentran en la etapa de "Obra Terminada" (2,891) y "Obra en Ejecución" (2,218). Otras regiones como Maule, Ñuble y Los Lagos también muestran una presencia significativa de proyectos, destacando por su cantidad en "Obra Terminada". Por otro lado, regiones como Arica y Magallanes tienen una presencia más limitada en la base de datos, con un número considerablemente menor de proyectos, y en algunos casos, los proyectos se encuentran en la etapa de "Proyecto en Evaluación" o "Sin Proyecto Ingresado."

Este análisis detallado de la distribución de proyectos según el estado y la región de ejecución brinda información valiosa para comprender cómo están distribuidos geográficamente y en qué etapas se encuentran. Además, se observa que la categoría "Eliminado/Renunciado" está presente en varias regiones, lo que podría requerir una evaluación adicional para comprender las razones detrás de estos proyectos en ese estado

In [ ]:
pd.crosstab(df_f['Región Ejecución'],df_f['ESTADO PROYECTO'],margins=True)

ESTADO PROYECTO   ELIMINADO/RENUNCIADO  OBRA EN EJECUCIÓN  OBRA TERMINADA  \
Región Ejecución                                                            
Antofagasta                         14                  0              23   
Araucanía                          289               2218            2891   
Arica                                0                  0               0   
Atacama                             16                128             114   
Aysén                               98                185             337   
Biobío                              52                829             338   
Coquimbo                            98                241             589   
Los Lagos                          138                471            1222   
Los Ríos                            85                733            1053   
Magallanes                           5                  0              15   
Maule                              326               1938            2895   
Metropolitana                      187                381             903   
O'higgins                           87                475             301   
Tarapacá                             1                  3              30   
Valparaíso                         173                556             613   
Ñuble                               42               3091             893   
All                               1611              11249           12217   

ESTADO PROYECTO   PROYECTO APROBADO  PROYECTO EN EVALUACIÓN  \
Región Ejecución                                              
Antofagasta                      34                       0   
Araucanía                      2141                    1880   
Arica                             0                      34   
Atacama                           1                       1   
Aysén                           118                      60   
Biobío                          395                     418   
Coquimbo                        148                     437   
Los Lagos                       734                     384   
Los Ríos                        358                     855   
Magallanes                        0                      73   
Maule                           662                     450   
Metropolitana                   432                     104   
O'higgins                       121                     449   
Tarapacá                          0                      21   
Valparaíso                      310                     244   
Ñuble                          1352                     228   
All                            6806                    5638   

ESTADO PROYECTO   SIN PROYECTO INGRESADO    All  
Región Ejecución                                 
Antofagasta                            4     75  
Araucanía                           1802  11221  
Arica                                  0     34  
Atacama                               16    276  
Aysén                                  5    803  
Biobío                               550   2582  
Coquimbo                             412   1925  
Los Lagos                            442   3391  
Los Ríos                             264   3348  
Magallanes                             0     93  
Maule                               1396   7667  
Metropolitana                        166   2173  
O'higgins                            767   2200  
Tarapacá                              37     92  
Valparaíso                           213   2109  
Ñuble                                669   6275  
All                                 6743  44264

## Propuesta de Visualización

A fin de dar respuesta a las preguntas plateadas y con los datos disponibles se plantea el uso de las siguientes visualizaciones:

1. Un gráfico de segmentación puede ayudar a visualizar cómo se divide el programa en diferentes como los componentes del Programa (Vivienda Nueva, Mejoramiento y/o Ampliación de la Vivienda Existente) o sus Tipologías (Construcción de Conjunto Habitacional, Construcción en Sitio del Residente, Mejoramiento de Vivienda Existente, Ampliación de Vivienda Existente entre otros) Puedes asignar cada componente del programa a una sección del gráfico el Estado de las Obras, por año de termino cada uno. Esto permitirá analizar y comparar los componentes en términos de importancia y eficiencia.

2. Un mapa de calor puede ser útil para identificar cuellos de botella o puntos críticos en términos de ejecución territorial. Puedes utilizar un mapa de calor para representar la densidad de proyectos o actividades en diferentes áreas geográficas. Las áreas más congestionadas o con problemas de ejecución aparecerán en colores más intensos, lo que facilita la identificación de los puntos críticos.

3. Un mapa de clasificación (también conocido como choropleth map) puede ser útil para identificar territorios a nivel de comunas o región, más desfavorecidos en la ejecución de proyectos. Asignando colores diferentes a las áreas geográficas según su desempeño. Las áreas con evaluaciones positivas se destacarán con colores más brillantes, mientras que las áreas desfavorecidas se destacarán con colores más oscuros.
